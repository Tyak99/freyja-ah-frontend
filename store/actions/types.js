export const GET_ERRORS = 'GET_ERRORS';

export const SET_CURRENT_USER = 'SET_CURRENT_USER';

export const INIT_AUTH_REQUEST = 'INIT_AUTH_REQUEST';

export const END_AUTH_REQUEST = 'END_AUTH_REQUEST';

export const SIGNUP_ERROR = 'SIGNUP_ERROR';

export const LOGIN_ERROR = 'LOGIN_ERROR';
