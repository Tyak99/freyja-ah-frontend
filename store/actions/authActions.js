/* eslint-disable arrow-parens */
/* eslint-disable implicit-arrow-linebreak */
/* eslint-disable import/prefer-default-export */
import axios from 'axios';
import jwtDecode from 'jwt-decode';
import { GET_ERRORS, SET_CURRENT_USER } from './types';

// Set loged in user
export const setCurrentUser = decoded => ({
  type: SET_CURRENT_USER,
  payload: decoded,
});
export const setError = error => ({
  type: GET_ERRORS,
  payload: error,
});

// Register
export const registerUser = (newUser, history) => dispatch => {
  return axios
    .post('https://freyja-ah-backend.herokuapp.com/api/users', newUser)
    .then(res => {
      // Save to local storage
      const { token } = res.data;
      // Decode token to get user data
      const decoded = jwtDecode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
      history.push('/modules');
    })
    .catch(
      error => {
        dispatch(setError(error));
        console.log(error)
      },

      // eslint-disable-next-line function-paren-newline
    );
};
