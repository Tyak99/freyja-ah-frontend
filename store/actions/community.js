import { get } from 'axios';
import { GET_COMMUNITY_MESSAGES, GET_COMMUNITY_MESSAGES_ERROR } from './types';

export const getCommunityMessages = () => async (dispatch) => {
  try {
    const response = await get('https://freyja-ah-backend.herokuapp.com/api/community/messages', {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });
    console.log(response);
  } catch (error) {
    console.log(error.response);
  }
};
