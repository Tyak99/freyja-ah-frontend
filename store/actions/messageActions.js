import axios from 'axios';
import { GET_MESSAGE_SUCCESS, UPDATE_MESSAGE_LIST } from './types';

const url = 'http://localhost:3000/api/messages/';
export const getMessageSuccess = messages => ({
  type: GET_MESSAGE_SUCCESS,
  messages,
});

export const updateMessages = data => ({
  type: UPDATE_MESSAGE_LIST,
  data,
});

export const getMessage = () => (dispatch) => {
  axios.get(url, {
    headers: { Authorization: `${localStorage.getItem('token')}` },
  })
    .then((res) => {
      console.log('TCL: getMessage -> res', res.data.data);
      dispatch(getMessageSuccess(res.data.data));
    })
    .catch(error => console.log(error));
};
