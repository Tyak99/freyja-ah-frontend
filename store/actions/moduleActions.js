import { get } from 'axios';
import {
  GET_ALL_MODULES, GET_SINGLE_MODULE, MODULES_ERROR,
} from './types';

const baseUrl = 'https://freyja-ah-backend.herokuapp.com/api/';

const getModules = () => async (dispatch) => {
  try {
    const modulesRequest = await get('https://freyja-ah-backend.herokuapp.com/api/modules', {
      headers: { Authorization: `${localStorage.getItem('token')}` },
    });

    const { data } = modulesRequest.data;
    return dispatch({
      type: GET_ALL_MODULES,
      payload: data,
    });
  } catch (error) {
    return dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};


export const setModule = payload => ({
  type: GET_SINGLE_MODULE,
  payload,
});

export const getModule = moduleId => async (dispatch) => {
  try {
    const request = await get(`${baseUrl}/modules/${moduleId}`);
    const { data } = request.data;
    dispatch(setModule(data));
  } catch (error) {
    console.log(error.response);
    dispatch({
      type: MODULES_ERROR,
      payload: error,
    });
  }
};

export { getModules };
