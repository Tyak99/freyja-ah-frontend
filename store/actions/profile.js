import axios from 'axios';
import * as actionCreators from '../constants/profileContants';

const url = 'https://freyja-ah-backend.herokuapp.com/api/profiles';
const token = localStorage.getItem('jwtToken');
export const getProfileSuccess = profile => ({
  type: actionCreators.GET_PROFILE_SUCCESS,
  data: profile,
});

export const getProfileFailed = error => ({
  type: actionCreators.GET_PROFILE_FAILED,
  error,
});
export const getProfile = userId => async (dispatch) => {
  const config = {
    headers: {
      Authorization: token,
    },
  };
  try {
    const profile = await axios.get(`${url}/${userId}`, config);
    const { data } = profile;
    const { user } = data;

    const userDetails = {
      name: `${user.firstName} ${user.lastName}`,
      email: user.email,
      bio: user.profile.bio,
      image: user.profile.image || 'https://cdn.umnbootcamp.com/wp-content/uploads/sites/93/2018/03/placeholder-person.png',
      username: user.userName,
      phoneNumber: user.profile.phoneNumber,
      socialMedia: [
        { name: 'facebook', url: user.profile.facebook },
        { name: 'twitter', url: user.profile.twitter },
        { name: 'linkedin', url: user.profile.linkedIn },
      ],
      DOB: user.profile.dateOfBirth,
      industry: user.profile.industry,
      yrsOfExperience: user.profile.yrsOfExperience,
    };

    dispatch(getProfileSuccess(userDetails));
  } catch (error) {
    dispatch(getProfileFailed(error));
  }
};
