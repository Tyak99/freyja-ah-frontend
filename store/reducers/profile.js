import * as actionTypes from '../constants/profileContants';

export const initialState = {
  error: null,
  data: {},
};

export default (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.GET_PROFILE_SUCCESS:
      return {
        ...state,
        data: action.data,
      };
    case actionTypes.GET_PROFILE_FAILED:
      return {
        ...state,
        error: action.error,
      };
    default:
      return state;
  }
};
