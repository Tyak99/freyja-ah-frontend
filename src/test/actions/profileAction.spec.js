import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import { getProfile, getProfileSuccess } from '../../../store/actions/profile';
import profileReducer, { initialState } from '../../../store/reducers/profile';

const token = 'token';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const mockProfile = {
  data: {
    id: 1,
    phoneNumber: 3223232,
    dateOfBirth: '12/12/12',
    isEmployed: true,
    yrsOfExperience: 4,
    industry: 'Andela',
    image: 'http://',
    isCertified: false,
  },
};

describe('action creators', () => {
  const store = mockStore({});

  it('calls PROFILE_SUCCESS when getProfile is successful', (done) => {
    store.dispatch(getProfileSuccess(mockProfile));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
});
describe('async actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    // clear all HTTP mocks after each test
    nock.cleanAll();
  });
  it('should create getProfileSuccess when profile has been gotten', (done) => {
    nock('https://freyja-ah-backend.herokuapp.com')
      .get('/api/profiles/4') // Route to catch and mock
      .reply(200, {
        user: {
          userId: 4,
          id: 1,
          firstName: 'John',
          lastName: 'Grisham',
          userName: 'Tyak',
          profile: {
            age: 40,
            phoneNumber: 7689567890,
            bio: 'Help me',
          },
        },
      });
    return store.dispatch(getProfile(4, token)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
  it('should dispatch getProfileFailed when getProfile fails', (done) => {
    nock('https://freyja-ah-backend.herokuapp.com')
      .get('/api/profiles/99') // Route to catch and mock
      .reply(404, { error: 'user does not exist' });
    return store.dispatch(getProfile(4, token)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
});

describe('profile reducer', () => {
  it('should test the initial state', () => {
    expect(profileReducer(undefined, {})).toMatchSnapshot();
  });
  it('should test the state for update when get profile fails', () => {
    expect(profileReducer(initialState, { type: 'GET_PROFILE_FAILED' })).toMatchSnapshot();
  });
  it('should test the state for update when get profile success', () => {
    expect(profileReducer(initialState, { type: 'GET_PROFILE_SUCCESS' })).toMatchSnapshot();
  });
});
