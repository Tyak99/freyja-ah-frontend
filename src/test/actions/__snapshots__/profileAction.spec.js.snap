// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators calls PROFILE_SUCCESS when getProfile is successful 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "dateOfBirth": "12/12/12",
        "id": 1,
        "image": "http://",
        "industry": "Andela",
        "isCertified": false,
        "isEmployed": true,
        "phoneNumber": 3223232,
        "yrsOfExperience": 4,
      },
    },
    "type": "GET_PROFILE_SUCCESS",
  },
]
`;

exports[`async actions should create getProfileSuccess when profile has been gotten 1`] = `
Array [
  Object {
    "data": Object {
      "DOB": undefined,
      "bio": "Help me",
      "email": undefined,
      "image": "https://cdn.umnbootcamp.com/wp-content/uploads/sites/93/2018/03/placeholder-person.png",
      "industry": undefined,
      "name": "John Grisham",
      "phoneNumber": 7689567890,
      "socialMedia": Array [
        Object {
          "name": "facebook",
          "url": undefined,
        },
        Object {
          "name": "twitter",
          "url": undefined,
        },
        Object {
          "name": "linkedin",
          "url": undefined,
        },
      ],
      "username": "Tyak",
      "yrsOfExperience": undefined,
    },
    "type": "GET_PROFILE_SUCCESS",
  },
]
`;

exports[`async actions should dispatch getProfileFailed when getProfile fails 1`] = `
Array [
  Object {
    "error": [Error: Nock: No match for request {
  "method": "GET",
  "url": "https://freyja-ah-backend.herokuapp.com/api/profiles/4",
  "headers": {
    "accept": "application/json, text/plain, */*",
    "authorization": {},
    "user-agent": "axios/0.19.0"
  }
}],
    "type": "GET_PROFILE_FAILED",
  },
]
`;

exports[`profile reducer should test the initial state 1`] = `
Object {
  "data": Object {},
  "error": null,
}
`;

exports[`profile reducer should test the state for update when get profile fails 1`] = `
Object {
  "data": Object {},
  "error": undefined,
}
`;

exports[`profile reducer should test the state for update when get profile success 1`] = `
Object {
  "data": undefined,
  "error": null,
}
`;
